File: Sean的leetcode/Leetcode/Evaluate Division.cpp
class Solution {
public:
    vector<double> calcEquation(vector<pair<string, string>> equations, vector<double>& values, vector<pair<string, string>> queries) {
        unordered_map<string, unordered_map<string, double>> table;
        for (int i = 0; i < equations.size(); i++) {
            table[equations[i].first][equations[i].second] = values[i];
            table[equations[i].second][equations[i].first] = 1.0 / values[i];
        }
        vector<double> res(queries.size());
        
        int i = 0;
        for (auto q : queries) {
            unordered_set<string> visit;
            double ans = find_res(q.first, q.second, table, visit);
            res[i++] = ans == NULL? -1.0 : ans;
        }
        return res;
    }
    
    double find_res(string num, string deno, unordered_map<string, unordered_map<string,double>>& table, unordered_set<string>& visit) {
        string key = num + ":" + deno;
        if (visit.count(key)) return NULL;
        visit.insert(key);
        if (!table.count(num) || !table.count(deno)) return NULL;
        if (num == deno) return 1.0;
        
        unordered_map<string, double> paths = table[num];
        if (table[num].count(deno)) {
            return table[num][deno];
        }
        for (auto p : paths) {
            double ans = find_res(p.first, deno, table, visit);
            if (ans != NULL){
                return p.second * ans;
            } 
        }
        visit.erase(key);
        return NULL;
    
        
    }
    
};

File: Sean的leetcode/python/Evaluate Division.py
###
# Graph
# DFS
# Time Complexity: O(E) (build a graph) + O(Q*E) 
###
class Solution(object):
    def calcEquation(self, equations, values, queries):
        """
        :type equations: List[List[str]]
        :type values: List[float]
        :type queries: List[List[str]]
        :rtype: List[float]
        """
        from collections import defaultdict
        d = defaultdict(dict)
        for (nu, de), v in zip(equations, values):
            d[nu][de] = v
            d[de][nu] = 1.0 / v
        re = []
        for nu, de in queries:
            ans = self.dfs(nu, de, d, set())
            if not ans:
                re.append(-1.0)
            else:
                re.append(ans)    
        return re
    
    def dfs(self, nu, de, d, visit):
        if nu in visit:
            return None
        if nu not in d or de not in d:
            return None
        if nu == de:
            return 1.0
        if de in d[nu]:
            return d[nu][de]
        visit.add(nu)
        for e in d[nu].keys():
            res = self.dfs(e, de, d, visit)
            if res:
                return res * d[nu][e]
        visit.remove(nu)
        return None
        
        